### PR TITLE
New Crash Screen

### PR DESCRIPTION
**Pull request by gheskett**
_TODO: In-depth description still needed._

### CODE REVIEW

File: include/sm64.h [hunk 1]
```diff
@@ -15,11 +15,21 @@
...
 
-// Crash handler enhancement
-#ifdef CRASH_SCREEN_INCLUDED
-#define DEBUG_ASSERT(exp) do { if (!(exp)) _n64_assert(__FILE__, __LINE__, #exp, 1); } while (0)
+extern char *__n64Assert_Filename;
+extern u32   __n64Assert_LineNum;
+extern char *__n64Assert_Message;
+extern void __n64Assert(char *fileName, u32 lineNum, char *message);
+
+#define ASSERT(cond, message) do {\
```
Comment (someone2639) [inline include/sm64.h:23]: imo if this is a runtime assert that follows the C standard then it should be named `assert` in all lowercase

File: src/crash_screen/crash_screen.c [hunk 2]
```diff
@@ -0,0 +1,1569 @@
...
+    /*FPCSR_CE*/ "Unimplemented operation",
+    /*FPCSR_CV*/ "Invalid operation",
+    /*FPCSR_CZ*/ "Division by zero",
+    /*FPCSR_CO*/ "Overflow",
+    /*FPCSR_CU*/ "Underflow",
+    /*FPCSR_CI*/ "Inexact operation",
+};
+
+static const char *sRegNames[29] = {
+    "AT", "V0", "V1",
```
Comment (someone2639) [inline src/crash_screen/crash_screen.c:88]: what happened to `r0` lol (also theres 32 general registers where did the other 2 besides `r0` go)

File: src/crash_screen/insn_disasm.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    PARAM_SYN, //
+    PARAM_S,   // rs
+    PARAM_T,   // rt
+    PARAM_D,   // rd
+    PARAM_ST,  // rs, rt
+    PARAM_ST2, // rs, rt
+    PARAM_DS,  // rd, rs
+    PARAM_TD,  // rt, rd
+    PARAM_SD,  // rs, rd
+    PARAM_STD, // rs, rd, rt
```
Comment (someone2639) [inline src/crash_screen/insn_disasm.h:21]: SDT?

File: src/crash_screen/map_parser.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+extern u8 _mainSegmentStart[];
+extern u8 _mainSegmentTextEnd[];
+extern u8 _engineSegmentStart[];
+extern u8 _engineSegmentTextEnd[];
+extern u8 _goddardSegmentStart[];
+extern u8 _goddardSegmentTextEnd[];
+
+s32 is_in_code_segment(uintptr_t addr) {
```
Comment (someone2639) [inline src/crash_screen/map_parser.c:43]: if `boot` segment doesnt get overwritten it should be added here

File: src/crash_screen/map_parser.h [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <ultra64.h>
+
+#include "types.h"
+
+
+struct MapEntry {
+    /*0x00*/ uintptr_t addr;
+    /*0x04*/ size_t name_offset;
```
Comment (someone2639) [inline src/crash_screen/map_parser.h:10]: should be uintptr_t since its an offset

File: src/crash_screen/map_parser.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    /*0x08*/ size_t name_len;
+    /*0x0C*/ size_t pad;
+}; /*0x10*/
+
+#define ADDR_IS_KNOWN(addr) ((*(uintptr_t*)(addr) & 0x80000000) != 0)
```
Comment (someone2639) [inline src/crash_screen/map_parser.h:15]: this could probably be merged into `IS_IN_SEGMENT` with no issues

File: include/sm64.h (hunk 1 above)
Comment (someone2639) [reply include/sm64.h:23]: on second thought that requires a bit of extra work (including meddling with n64_assert.s and the assert parser) so consider this comment optional

**Review by someone2639 — CHANGES_REQUESTED** [review]

**Comment by someone2639** [comment]
Something I would like to see on this branch is a list of things this PR will do, so that scope is limited to only those things and I can gauge what parts I can contribute to (I'm willing to write the list but I do want mutual agreement here).

**Comment by Arceveti** [comment]
Moving to new PR

**Pull request closed without merging**